src/main/java/org/truffleruby/core/numeric/FixnumNodes.java

```diff
@@ -1329,13 +1332,20 @@ public long powTwoLong(int a, int b) {
         }, limit = "getLimit()")
         public Object powConstantExponent(Object base, int exponent,
                 @Cached("exponent") int cachedExponent,
+                @Cached("create()") BranchProfile overflowProfile,
                 @Cached("create()") IntegerMulNode mulNode) {
             Object result = 1;
             int exp = cachedExponent;
             while (exp > 0) {
                 if ((exp & 1) == 0) {
                     base = mulNode.executeMul(base, base);
                     exp >>= 1;
+
+                    if (base instanceof DynamicObject) {
+                        overflowProfile.enter();
+                        final Object bignumResult = bignumPow(base, exp);
```

Comment:
Should this just be
`return bignumPow(originalBase, exponent)`?
Otherwise the exponent too big check might not trigger, isn't it?

Reply:
I basically copied MRI here. The one change I made is MRI checks if the value would overflow first and then promotes the current value of `base` to a `Bignum` via `rb_int2big`. We don't have a nice way to promote primitives to `Bignum`, which would be necessary to call `bignumPow`. So, I waited until the overflow, at which point we should have a `Bignum` and call it at that point.

To rephrase, I'm doing basically what MRI does in both `Fixnum#**` and `Bignum#**`. Because `Bignum#**` considers both arguments in its overflow check, I think we're fine.

Reply:
The check is basically `bitLength(base) * exp > limit`, in this loop we divide `exp` by 2 and multiply `base` by itself, so  `bitLength(base*base) should be around 2*bitLength(base)` and it's more or less the same condition.
There might be some corner cases though.
I'll let you decide. Please add a comment if you leave it like this.
We could extract some methods in the Bignum#** node that takes BigInteger so we could pass the original value if we want to make sure.

Reply:
I see, so the covered range wouldn't be exactly the same as MRI's. Although in your original proposal, it also wouldn't be. I don't know how much any of this matters in practice, but I can update this match MRI exactly.